README.md
# Thermal Vision

[![hacs][hacsbadge]][hacs]
[![GitHub Release][releases-shield]][releases]
[![GitHub Activity][commits-shield]][commits]
[![License][license-shield]](LICENSE)

The Thermal Vision integration allows for the use of the [AMG8833 sensor](https://eu.industrial.panasonic.com/products/sensors-optical-devices/sensors-automotive-and-industrial-applications/infrared-array/series/grid-eye-high-performance-type-amg8833/ADI8005/model/AMG8833) with a simple JSON interface to be used within [Home Assistant](https://www.home-assistant.io/).

This can be used to detect humans much more reliably than motion sensors as it does not require the subject to be moving.  It also produces a pretty thermal camera!

Firmware for a simple sensor using an ESP8266 (I'm using a D1 Mini) can built using [TheRealWaldo/esp8266-amg8833](https://github.com/TheRealWaldo/esp8266-amg8833).  This firmware is still under development!

Also, there's an experimental implementation using [ESPHome](https://esphome.io/) at [TheRealWaldo/AMG8833-ESPHOME](https://github.com/TheRealWaldo/AMG8833-ESPHOME).  You can leverage this by setting `pixel_sensor` to the sensor created by ESPHome and not using `host`.  Note: this only works with the camera at this time.

![Me, waving](docs/waving.png)

This is a rework of [eyalcha/thermal](https://github.com/eyalcha/thermal) that corrects a number of issues, makes it compatible with HA 2016.6 forward, and adds new functionality.

## Differences

This is may not be a complete list:

- The platform/domain name is `thermal_vision` instead of `thermal`
- More resilient connections
- Fixes issues:
  - https://github.com/eyalcha/thermal/issues/1
  - https://github.com/eyalcha/thermal/issues/2
  - https://github.com/eyalcha/thermal/issues/5
  - https://github.com/eyalcha/thermal/issues/6
- Works with HA 2016.6 and up
- Allows for both imperial and metric measurements
- Optional auto-ranging capabilities
- Optional in-camera overlay of metrics

## Installation

There is three options to install this integration; HACS is the easiest.

## INSTALLATION VIA HACS

Just go into the HACS Integrations section, search for the "Thermal Vision" repository, then click Install!

### MANUAL INSTALLATION VIA HACS

Follow the instructions at https://hacs.xyz/docs/faq/custom_repositories/ using the URL https://github.com/TheRealWaldo/thermal

### MANUAL INSTALLATION

1. Download the zip file from
   [latest release](https://github.com/TheRealWaldo/thermal/releases/latest).
2. Unpack the release and copy the `custom_components/thermal_vision` directory
   into the `custom_components` directory of your Home Assistant
   installation.
3. Configure the `thermal_vision` sensor and/or camera.
4. Restart Home Assistant.

## Configuration

### Camera

```yaml
# Example configuration.yaml entry

camera:
  - platform: thermal_vision
    host: http://192.168.0.10
```

#### Main Options

|Parameter |Required|Description
|:---|---|---
| `platform` | Yes | Platform name `thermal_vision`
| `name` | No | Friendly name **Default**: `Thermal Vision`
| `host` | Yes | IP address or hostname of your Thermal sensor server
| `pixel_sensor` | Required if `host` is not set; sensor containing the base64 encoded pixels
| `verify_ssl` | No | Verify SSL or not **Default**: `false`
| `width` | No | Image width in pixels **Default**: `640`
| `height` | No | Image height in pixels **Default**: `640`
| `preserve_aspect_ratio` | No | Preserve aspect ratio (ignores height) **Default**: `true`
| `rotate` | No | Rotate image **Default**: `0`
| `mirror` | No | Mirror image true / false **Default**: `false`
| `format` | No | Camera image format (`jpeg`, `png`) **Default**: `jpeg`
| `min_temp` | No | Min temperature **Default**: `26`
| `max_temp` | No | Max temperature **Default**: `32`
| `auto_range` | No | Rather than use a static minimum and maximum temperature, auto adjust based on the content ***Default***: `false`
| `min_diff` | No | Minimum difference when auto-ranging.  Favors cold.  ***Default***: `4`
| `sensor` | No | Sensor related configurations (see below)
| `interpolate` | No | Interpolation related configurations (see below)
| `cold_color` | No | Cold color **Default**: `indigo`
| `hot_color` | No | Hot color **Default**: `red`
| `session_timeout` | No | Timeout in seconds for polling Thermal sensor server **Default**: `2`
| `overlay` | No | Add an overlay to the image to visualize min/max temperature **Default**: `false`

Interpolate

|Parameter |Required|Description
|:---|---|---
| `method` | No | Interpolation method (`bicubic`, `linear`) **Default**: `bicubic`
| `rows` | No | Number of rows in interpolated data **Default**: `32`
| `cols` | No | Number of columns of interpolated data **Default**: `32`

Sensor

|Parameter |Required|Description
|:---|---|---
| `rows` | No | Number of rows in sensor data **Default**: `8`
| `cols` | No | Number of columns in sensor data **Default**: `8`

#### State and Attributes

##### Attributes

|Attribute |Description
|:---|---
| `fps` | Approximate frames per second based on response time of sensor
| `min` | Minimum value represented by the `cold_color` in the image
| `max` | Maximum value represented by the `hot_color` in the image

### Sensor

```yaml
# Example configuration.yaml entry

sensor:
  - platform: thermal_vision
    host: http://192.168.0.10
```

#### Main Options

|Parameter |Required|Description
|:---|---|---
| `platform` | Yes | Platform name `thermal_vision`
| `name` | No | Friendly name **Default**: `Thermal Vision`
| `host` | Yes | IP address of your Thermal sensor server
| `verify_ssl` | No | Verify SSL or not **Default**: `false`
| `scan_interval` | No | Get raw data interval in seconds **Default**: `60`
| `sensor` | No | Sensor related configurations (see below)
| `roi` | No | Sensor region of interest (see below)
| `state` | No | Sensor state type (`average`, `max`, `min`, `sensor_temp`, `person_detected` (*deprecated, use `binary_sensor`*)) **Default**: `average`

Sensor

|Parameter |Required|Description
|:---|---|---
| `rows` | No | Number of rows in sensor data **Default**: `8`
| `cols` | No | Number of columns in sensor data **Default**: `8`

ROI

|Parameter |Required|Description
|:---|---|---
| `left` | No | Left pixel index [0:cols-1] **Default**: `0`
| `top` | No | Top pixel index [0:rows-1] **Default**: `0`
| `right` | No | Right pixel index [0:cols-1] **Default**: `7`
| `bottom` | No | Bottom pixel index [0:rows-1] **Default**: `7`

#### State and Attributes

##### State

The sensor state can be either `average`, `max`, `min`, or `person_detected` based on `state` in the configuration.  It defaults to `average`.

##### Attributes

All values are affected by the ROI configuration.

|Attribute |Description
|:---|---
| `average` | Average temperature of all pixels in current capture
| `max` | Maximum temperature of all pixels in current capture
| `min` | Min temperature of all pixels in current capture
| `min_index` | The index where the min temperature was detected (1 Dimensional)
| `max_index` | The index where the max temperature was detected (1 Dimensional)
| `sensor_temp` | The temperature of the sensor itself (if the sensor provides it)
| `person_detected` | A boolean representing whether the sensor detected a person or not.  Must use latest firmware! (*deprecated, use `binary_sensor`*)

### Binary Sensor

```yaml
# Example configuration.yaml entry

binary_sensor:
  - platform: thermal_vision
    host: http://192.168.0.10
```

#### Main Options

|Parameter |Required|Description
|:---|---|---
| `platform` | Yes | Platform name `thermal_vision`
| `name` | No | Friendly name **Default**: `Thermal Vision`
| `host` | Yes | IP address of your Thermal sensor server
| `verify_ssl` | No | Verify SSL or not **Default**: `false`
| `scan_interval` | No | Get raw data interval in seconds **Default**: `60`

#### State and Attributes

##### State

Returns `on` when the sensor claims to detect a person, `off` when it does not.  Device class is `occupancy`.

[commits]: https://github.com/TheRealWaldo/thermal/commits/main
[commits-shield]: https://img.shields.io/github/commit-activity/m/therealwaldo/thermal?style=for-the-badge
[license-shield]: https://img.shields.io/github/license/therealwaldo/thermal.svg?style=for-the-badge
[hacs]: https://github.com/custom-components/hacs
[hacsbadge]: https://img.shields.io/badge/HACS-Default-orange.svg?style=for-the-badge
[releases-shield]: https://img.shields.io/github/v/release/therealwaldo/thermal?include_prereleases&style=for-the-badge
[releases]: https://github.com/TheRealWaldo/thermal/releases


CHANGELOG.md
# Changelog

## [3.0.0](https://www.github.com/TheRealWaldo/thermal/compare/v2.6.0...v3.0.0) (2021-09-08)


### ⚠ BREAKING CHANGES

* Using `person_detected` as a state on a sensor is deprecated and will be removed in future versions.  Use the new Binary Sensor instead.

### Features

* add state_class to sensor ([8019b86](https://www.github.com/TheRealWaldo/thermal/commit/8019b869a7d7f03246e85cc0cf4cb2814fe0da61))
* introduce binary_sensor for occupancy ([f0ee349](https://www.github.com/TheRealWaldo/thermal/commit/f0ee349c4d0f92c42929d4589f9b0121ce02c097))
* use hass native unit of measurement ([247487d](https://www.github.com/TheRealWaldo/thermal/commit/247487d7a873130a5a659d343e4279d53d219c3b))

## [2.6.0](https://www.github.com/TheRealWaldo/thermal/compare/v2.5.0...v2.6.0) (2021-09-07)


### Features

* support scaling, preserve aspect ratio ([0d02ea0](https://www.github.com/TheRealWaldo/thermal/commit/0d02ea05dc0f46d8db0a488d703bf7776b843add))

## [2.5.0](https://www.github.com/TheRealWaldo/thermal/compare/v2.4.2...v2.5.0) (2021-09-07)


### Features

* allow for min_diff when auto_range enabled ([f2a2e3c](https://www.github.com/TheRealWaldo/thermal/commit/f2a2e3c07891eb5a708f15c8140310b7b560d2b5))


### Documentation

* update HACS install steps ([967791a](https://www.github.com/TheRealWaldo/thermal/commit/967791add722b99768d59e79626b91aa867c7a70))

### [2.4.2](https://www.github.com/TheRealWaldo/thermal/compare/v2.4.1...v2.4.2) (2021-08-26)


### Bug Fixes

* person_detected and sensor_temp state ([78b4984](https://www.github.com/TheRealWaldo/thermal/commit/78b498419da6559cca9bc4d7d21b54fe8034935d))

### [2.4.1](https://www.github.com/TheRealWaldo/thermal/compare/v2.4.0...v2.4.1) (2021-08-26)


### Bug Fixes

* sensor would not start ([3402fdc](https://www.github.com/TheRealWaldo/thermal/commit/3402fdcec5cb50d21a2f2dc514961492dcd18b22))

## [2.4.0](https://www.github.com/TheRealWaldo/thermal/compare/v2.3.0...v2.4.0) (2021-08-19)


### Features

* add rudimentary person_detected support ([1dfa585](https://www.github.com/TheRealWaldo/thermal/commit/1dfa585305c25125051b5d1444890aa9e7b01a0c))

## [2.3.0](https://www.github.com/TheRealWaldo/thermal/compare/v2.2.1...v2.3.0) (2021-08-18)


### Features

* add ability to leverage ESPHome based camera ([8b4a6d8](https://www.github.com/TheRealWaldo/thermal/commit/8b4a6d8ab6210784c3ca776f61e221dca09749b1))


custom_components/thermal_vision/const.py
"""Constants for Thermal Vision integration."""

DOMAIN = "thermal_vision"

VERSION = "3.0.0"
ISSUE_URL = "https://github.com/TheRealWaldo/thermal/issues"

CONF_WIDTH = "width"
CONF_HEIGHT = "height"
CONF_PRESERVE_ASPECT_RATIO = "preserve_aspect_ratio"
CONF_PIXEL_SENSOR = "pixel_sensor"
CONF_MIN_DIFFERANCE = "min_diff"
CONF_MIN_TEMPERATURE = "min_temp"
CONF_MAX_TEMPERATURE = "max_temp"
CONF_METHOD = "method"
CONF_AUTO_RANGE = "auto_range"
CONF_ROI = "roi"
CONF_TOP = "top"
CONF_LEFT = "left"
CONF_RIGHT = "right"
CONF_BOTTOM = "bottom"
CONF_STATE = "state"
CONF_SENSOR = "sensor"
CONF_ROWS = "rows"
CONF_COLS = "cols"
CONF_ROTATE = "rotate"
CONF_MIRROR = "mirror"
CONF_FORMAT = "format"
CONF_INTERPOLATE = "interpolate"
CONF_COLD_COLOR = "cold_color"
CONF_HOT_COLOR = "hot_color"
CONF_SESSION_TIMEOUT = "session_timeout"
CONF_OVERLAY = "overlay"

ATTR_MIN_INDEX = "min_index"
ATTR_MAX_INDEX = "max_index"
ATTR_MIN = "min"
ATTR_MAX = "max"
ATTR_AVG = "average"
ATTR_SENSOR_TEMP = "sensor_temp"
ATTR_PERSON_DETECTED = "person_detected"

DEFAULT_NAME = "Thermal Vision"
DEFAULT_VERIFY_SSL = False
DEFAULT_OVERLAY = False
DEFAULT_IMAGE_WIDTH = 640
DEFAULT_IMAGE_HEIGHT = 640
DEFAULT_PRESERVE_ASPECT_RATIO = True
DEFAULT_MIN_TEMPERATURE = 26.0
DEFAULT_MAX_TEMPERATURE = 32.0
DEFAULT_METHOD = "bicubic"
DEFAULT_STATE = ATTR_AVG
DEFAULT_ROWS = 8
DEFAULT_COLS = 8
DEFAULT_ROTATE = 0
DEFAULT_MIRROR = False
DEFAULT_COLD_COLOR = "indigo"
DEFAULT_HOT_COLOR = "red"
DEFAULT_FORMAT = "jpeg"
DEFAULT_SCAN_INTERVAL = 60
DEFAULT_SESSION_TIMEOUT = 2

SERVICE_AUTO_SCALE = "auto_scale"
